src/details/ArborX_DetailsNearestBufferProvider.hpp

```diff
@@ -0,0 +1,73 @@
+/****************************************************************************
+ * Copyright (c) 2017-2023 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_DETAILS_NEAREST_BUFFER_PROVIDER_HPP
+#define ARBORX_DETAILS_NEAREST_BUFFER_PROVIDER_HPP
+
+#include <ArborX_DetailsKokkosExtViewHelpers.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX::Details
+{
+
+template <typename MemorySpace>
+struct NearestBufferProvider
+{
+  static_assert(Kokkos::is_memory_space_v<MemorySpace>);
+
+  using PairIndexDistance = Kokkos::pair<int, float>;
+
+  Kokkos::View<PairIndexDistance *, MemorySpace> _buffer;
+  Kokkos::View<int *, MemorySpace> _offset;
+
+  NearestBufferProvider() = default;
+
+  template <typename ExecutionSpace, typename Predicates>
+  NearestBufferProvider(ExecutionSpace const &space,
+                        Predicates const &predicates)
+      : _buffer("ArborX::NearestBufferProvider::buffer", 0)
+      , _offset("ArborX::NearestBufferProvider::offset", 0)
+  {
+    allocateBuffer(space, predicates);
+  }
+
+  KOKKOS_FUNCTION auto operator()(int i) const
+  {
+    auto const *offset_ptr = &_offset(i);
+    return Kokkos::subview(_buffer,
+                           Kokkos::make_pair(*offset_ptr, *(offset_ptr + 1)));
```

Comment:
Any good reason why we would do pointer arithmetic over regular view element accesses?

Reply:
That is a good question. This is inherited code. The earliest introduction I can see is in #282, but I can't tell you why this is done this way.

Reply:
Do you mind doing
```suggestion
                           Kokkos::make_pair(offset(i), offset(i + 1)));
```
and checking that there is no perf regression with it?

Reply:
Why would you expect a perf difference here? You essentially do the same thing. My only question about this code is whether I'm missing something about host/device dereferencing interaction.

Reply:
the code I suggest is better